### ISSUE TITLE
[red-knot] don't include Unknown in the type for a conditionally-defined import

### ISSUE THREAD

**Issue by carljm** [astral-sh/ruff#13514]
If `lib.py` has:
```
if flag:
    x: int = 1
```
and `main.py` has `from lib import x`, currently we infer the imported type of `x` as `Unknown | int`, with the `Unknown` coming from the "not bound" path. But this is silly; if the name is not bound, you'll get an error, but you won't get some random type. We should just infer the type of the import as `int` in this case (and possibly emit an "import may not be defined" diagnostic, but that's separate from this issue).

In other words, the unbound case should become `Never` (because if the name doesn't exist your code will stop running with an error), which will then disappear when unioned with the other path; it shouldn't become `Unknown`.

Both mypy and pyright also infer simply `int` here.

**Comment by Slyces** [comment]
I started to check out this issue, and I noticed the following inferences
```python
if flag:
    x: int = 1        # Unknown | int
    y = 1             # Unbound | Literal[1]
```
I had some questions:
- Is this difference in behaviour expected (given that `Literal[1]` is a specialisation of `int`, it probably shouldn't impact the `Unknown/Unbound` behaviour?)
- To fix the current issue, do we need to address both (`x` -> `int` & `y` -> `Literal[1]`)

**Comment by carljm** [comment]
> * Is this difference in behaviour expected

Yes and no -- it was expected but really neither of those types should show up in the public type (and `Unbound` really shouldn't be a type at all, it should be separate metadata about a particular use of a name), so it's not desired :)

> * do we need to address both (`x` -> `int` and `y` -> `Literal[1]`)

Yes, we should; sorry, that wasn't clear in my issue description.
